lib/worker/OperationQueue.js

```diff
@@ -0,0 +1,363 @@
+            if(!valid) {
+              this.removedOperations.push(operation);
+              // return early to prevent adding the operation
+              return;
+            }
+          }
+          // adds the first operation and all other successfully validated
+          // operations
+          operations.push(operation);
+        } catch(e) {
+          logger.debug(OPERATION_VALIDATION_ERROR,
+            {ledgerNodeId: this.ledgerNodeId, operation, error: e});
+          this.removedOperations.push(operation);
+        }
+      }));
+    }
+
+    await Promise.all(promises);
+
+    if(this.removedOperations.length > 0) {
+      const msg = 'Non-critical error. Removed invalid operations from chunk.';
+      const {ledgerNodeId} = this;
+      logger.debug(msg, {ledgerNodeId, n: this.removedOperations.length});
+    }
+
+    return operations;
+  }
+
+  async _getHead() {
+    let opCount = 0;
+    let head = null;
+
+    for(let i = this.head; i < this.queue.length; ++i) {
+      const operation = this.queue[i];
+      if(opCount >= this.MAX_OPERATIONS) {
+        break;
+      }
+      if(head === null) {
+        try {
+          const {valid} = await this._isValidOperation({operation});
+          if(!valid) {
+            this.removedOperations.push(operation);
+            logger.debug(OPERATION_VALIDATION_ERROR, {operation});
+            continue;
+          }
+          head = i;
+        } catch(e) {
+          this.removedOperations.push(operation);
+          logger.debug(OPERATION_VALIDATION_ERROR,
+            {ledgerNodeId: this.ledgerNodeId, operation, error: e});
+          continue;
+        }
+      }
+      ++opCount;
+    }
+
+    if(this.removedOperations.length > 0) {
+      this.removing = this._removeOperationsFromQueue().catch(e => {
+        logger.error(OPERATIONS_REMOVAL_ERROR,
+          {ledgerNodeId: this.ledgerNodeId, error: e});
+      });
+    }
+
+    return {head, opCount};
+  }
+
+  async _isDatabaseEmpty() {
+    const {ledgerNodeId} = this;
+    const projection = {_id: 1};
+    try {
+      const count = await this.collection.find({ledgerNodeId}, {projection})
+        .limit(1).count();
+      return count === 0;
+    } catch(e) {
+      logger.error('Error checking database for empty operation queue.', {
+        ledgerNodeId,
+        error: e
+      });
+      return true;
+    }
+  }
+
+  async _removeOperationsFromQueue({operations = []} = {}) {
+    for(const operation of operations) {
+      this.removedOperations.push(operation);
+    }
+
+    if(this.removedOperations.length === 0) {
+      return;
+    }
+
+    const {ledgerNodeId} = this;
+    try {
+      const bulkOp = this.collection.initializeUnorderedBulkOp();
+
+      for(const {meta} of this.removedOperations) {
+        bulkOp.find({
+          ledgerNodeId,
+          'meta.operationHash': meta.operationHash
+        }).delete();
+      }
+
+      await bulkOp.execute();
```

Comment:
I'm unfamiliar with this API -- going to trust you here.

Reply:
See: https://docs.mongodb.com/manual/reference/method/db.collection.initializeUnorderedBulkOp/